### PR TITLE
E UI 4628 order of display

### PR DESCRIPTION
**Pull request by waynepHMCTS**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/EUI-4700

### Change description ###
Order of hearings

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/cases/containers/case-hearings/case-hearings.component.ts [hunk 1]
```diff
@@ -53,6 +78,22 @@ export class CaseHearingsComponent implements OnInit, OnDestroy {
     }
   }
 
+  public convertServiceToVM(hearings: CaseHearingModel[]) {
```
Comment (rayliangatos) [inline src/cases/containers/case-hearings/case-hearings.component.ts:81]: Could we change the method name from 'convertServiceToVM' to 'calculateMostRecentHearingDate', as this is not a service.

File: src/hearings/models/caseHearing.model.ts [hunk 2]
```diff
@@ -1,11 +1,10 @@
 import { HearingDayScheduleModel } from './hearingDaySchedule.model';
-import { EXUIDisplayStatusEnum, EXUISectionStatusEnum } from './hearings.enum';
-
+import { EXUIDisplayStatusEnum, EXUISectionStatusEnum, HearingListingStatusEnum } from './hearings.enum';
 export interface CaseHearingModel {
   hearingID: string;
   hearingType: string;
   hmcStatus: string;
-  lastResponseReceivedDateTime: string;
```
Comment (rayliangatos) [inline src/hearings/models/caseHearing.model.ts:8]: lastResponseReceivedDateTime

**Review by rayliangatos — CHANGES_REQUESTED** [review]

File: src/cases/containers/case-hearings/case-hearings.component.ts [hunk 3]
```diff
@@ -39,8 +37,35 @@ export class CaseHearingsComponent implements OnInit, OnDestroy {
...
+      if (hearings.length) {
+        const viewModels: CaseHearingViewModel[] = this.convertServiceToVM(hearings);
+
+        this.upcomingHearings$ = of(viewModels.sort((a, b) => {
+          return new Date(a.hearingRequestDateTime) > new Date(b.hearingRequestDateTime) ? 1 : -1;
+        }).sort((a, b) => {
+          return new Date(a.mostRecentHearingStartDateTime) > new Date(b.mostRecentHearingStartDateTime) ? 1 : -1;
+        }).sort((a) => {
+          return a.hearingListingStatus === HearingListingStatusEnum.AWAITING_LISTING ? -1 : 1;
+        }));
```
Comment (rayliangatos) [inline src/cases/containers/case-hearings/case-hearings.component.ts:50]: according to requirement, should we order the mostRecentHearingStartDateTime first, if  mostRecentHearingStartDateTime  is not found(null) then order by hearingRequestDateTime?

Comment (waynepHMCTS) [reply src/cases/containers/case-hearings/case-hearings.component.ts:50]: This done the sort chaining is reverse order

File: src/hearings/containers/request-hearing/request-hearing.component.html [hunk 4]
```diff
@@ -1,11 +1 @@
...
-  <div class="page">
-    <main id="content" role="main">
-      <p>
-        request-hearing works!
-      </p>
-    </main>
-  </div>
-</div>
-<exui-app-footer></exui-app-footer>
+Request works!
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/request-hearing.component.html:1]: revert it plz

File: src/hearings/containers/request-hearing/request-hearing.component.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+.search-location {
+    width: 55%
+}
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/request-hearing.component.scss:3]: revert it plz

File: src/cases/components/case-hearings-list/case-hearings-list.component.spec.ts [hunk 6]
```diff
@@ -17,7 +17,7 @@ class MockRoleCategoryMappingService {
   public getValueOnce = <R>(key: string, defaultValue: R): Observable<R> => of(defaultValue);
 }
 
-describe('CaseHearingsListComponent', () => {
+fdescribe('CaseHearingsListComponent', () => {
```
Comment (rayliangatos) [inline src/cases/components/case-hearings-list/case-hearings-list.component.spec.ts:20]: fdescribe

File: api/hearings/models/caseHearing.model.ts [hunk 7]
```diff
@@ -4,8 +4,9 @@ import { EXUIDisplayStatusEnum, EXUISectionStatusEnum } from './hearings.enum';
 export interface CaseHearingModel {
   hearingID: string;
   hearingType: string;
-  hmcStatus: string;
+  hearingRequestDateTime: string,
```
Comment (rayliangatos) [inline api/hearings/models/caseHearing.model.ts:7]: ; please

File: src/cases/components/case-hearings-list/case-hearings-list.component.ts [hunk 8]
```diff
@@ -1,8 +1,7 @@
 import { Component, Input, OnInit } from '@angular/core';
 import { Observable } from 'rxjs';
-import { CaseHearingModel } from '../../../hearings/models/caseHearing.model';
+import { CaseHearingViewModel } from 'src/hearings/viewModel/case-hearing-view.model';
```
Comment (rayliangatos) [inline src/cases/components/case-hearings-list/case-hearings-list.component.ts:3]: src

File: src/hearings/viewModel/case-hearing-view.model.ts [hunk 9]
```diff
@@ -0,0 +1,4 @@
+import { CaseHearingModel } from 'api/hearings/models/caseHearing.model';
+export interface CaseHearingViewModel extends CaseHearingModel {
```
Comment (rayliangatos) [inline src/hearings/viewModel/case-hearing-view.model.ts:2]: Could this file put under existing models folder? Reason:
1. This is still a model
2. Doesn't make sensen to create a new folder just for 1 new file inside

**Review by rayliangatos — COMMENTED** [review]
new comments to be addressed

File: src/cases/containers/case-hearings/case-hearings.component.ts [hunk 10]
```diff
@@ -53,6 +76,44 @@ export class CaseHearingsComponent implements OnInit, OnDestroy {
...
+    hearings.forEach((hearing) => {
+      const viewModel = {} as CaseHearingViewModel;
+      Object.keys(hearing).forEach(key => viewModel[key] = hearing[key]);
+
+      if (hearing.hearingDaySchedule && hearing.hearingDaySchedule.length) {
+        const moments = hearing.hearingDaySchedule.map(d => moment(d.hearingStartDateTime ? d.hearingStartDateTime : this.minimumDate));
+        viewModel.mostRecentHearingStartDateTime = moment.max(moments).format('l');
+
+        if (viewModel.mostRecentHearingStartDateTime === 'Invalid date') {
+          viewModel.mostRecentHearingStartDateTime = this.minimumDate;
```
[the remark this answers is not in the document]
Comment (waynepHMCTS) [reply src/cases/containers/case-hearings/case-hearings.component.ts:96]: Moment returns Invalid date when mostRecent is set to nothing

[the remark this answers is not in the document]
Comment (rayliangatos) [reply src/cases/containers/case-hearings/case-hearings.component.ts:96]: If hearingDaySchedule: null or [], it doesn't need to calculate with the moment.max as the mostRecentHearingStartDateTime  can be set as null directly.

**Review by rayliangatos — APPROVED** [review]

**Review by Kartheek-S — APPROVED** [review]

**Pull request merged**